src/plugins/data/public/ui/dataset_navigator/dataset_navigator.tsx

```diff
@@ -0,0 +1,253 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useEffect, useState } from 'react';
+
+import { EuiButtonEmpty, EuiContextMenu, EuiPopover } from '@elastic/eui';
+import { SavedObjectsClientContract, SimpleSavedObject } from 'opensearch-dashboards/public';
+import { map, scan } from 'rxjs/operators';
+import { ISearchStart } from '../../search/types';
+import { IIndexPattern } from '../..';
+import { getUiService, getIndexPatterns, getSearchService } from '../../services';
+
+const getClusters = async (savedObjectsClient: SavedObjectsClientContract) => {
+  return await savedObjectsClient.find({
+    type: 'data-source',
+    perPage: 10000,
+  });
+};
+
+export const searchResponseToArray = (showAllIndices: boolean) => (response) => {
+  const { rawResponse } = response;
+  if (!rawResponse.aggregations) {
+    return [];
+  } else {
+    return rawResponse.aggregations.indices.buckets
+      .map((bucket: { key: string }) => {
+        return bucket.key;
+      })
+      .filter((indexName: string) => {
+        if (showAllIndices) {
+          return true;
+        } else {
+          return !indexName.startsWith('.');
+        }
+      })
+      .map((indexName: string) => {
+        return {
+          name: indexName,
+          // item: {},
+        };
+      });
+  }
+};
+
+const buildSearchRequest = (showAllIndices: boolean, pattern: string, dataSourceId?: string) => {
+  const request = {
+    params: {
+      ignoreUnavailable: true,
+      expand_wildcards: showAllIndices ? 'all' : 'open',
+      index: pattern,
+      body: {
+        size: 0, // no hits
+        aggs: {
+          indices: {
+            terms: {
+              field: '_index',
+              size: 100,
+            },
+          },
+        },
+      },
+    },
+    dataSourceId,
+  };
+
+  return request;
+};
+
+const getIndices = async (search: ISearchStart, dataSourceId: string) => {
+  const request = buildSearchRequest(true, '*', dataSourceId);
+  return search
+    .getDefaultSearchInterceptor()
+    .search(request)
+    .pipe(map(searchResponseToArray(true)))
+    .pipe(scan((accumulator = [], value) => accumulator.join(value)))
+    .toPromise()
+    .catch(() => []);
+};
+
+interface DataSetOption {
+  id: string;
+  name: string;
+  dataSourceRef?: string;
+}
+
+interface DataSetNavigatorProps {
+  savedObjectsClient: SavedObjectsClientContract;
+  indexPatterns: Array<IIndexPattern | string>;
+}
+
+export const DataSetNavigator = ({ savedObjectsClient, indexPatterns }: DataSetNavigatorProps) => {
+  const [isDataSetNavigatorOpen, setIsDataSetNavigatorOpen] = useState(false);
+  const [clusterList, setClusterList] = useState<SimpleSavedObject[]>([]);
+  const [indexList, setIndexList] = useState<DataSetOption[]>([]);
+  const [selectedCluster, setSelectedCluster] = useState<any>();
+  const [selectedDataSet, setSelectedDataSet] = useState<DataSetOption>({
+    id: indexPatterns[0]?.id,
+    name: indexPatterns[0]?.title,
```

Comment:
`indexPatterns[number]` can be a `string` based on its type, accessing `.id` and `.title` of a string will crash?